### PR TITLE
Prepare for Testing the ABI-Router in PyTeal

### PR DESCRIPTION
**Pull request by tzaffi**
## Summary of changes

* `class ABIMethodCallStrategy` (formerly `class ABIContractExecutor`):
  * This is moved into`graviton/abi_strategy.py` (from `graviton/blackbox.py`)
  * Removing the following methods entirely:
    * `validate_inputs()`
    * `dryrun_sequence()`
  * New methods:
    * `get()`
    * `num_args()`
* Introducing `class Simulation` in `graviton/sim.py`. See below for usage example.
* New test file `tests/integration/abi_router_test.py` which contains some tests formerly in `tests/integration/abi_test.py` but which are -as you might guess- all about testing the `Router`
* Fixing `DryRunInspector`'s `error()` and `error_message()` which recently broke

## PyTeal Sibling PR

https://github.com/algorand/pyteal/pull/634

## TODO

- [x] Timebox expired. Most likely [culprit](https://github.com/algorand/go-algorand/pull/4719). ~Figure out the actual PR referenced above and in the CHANGELOG~

### CODE REVIEW

File: graviton/blackbox.py [hunk 1]
```diff
@@ -542,183 +540,3 @@ def _abi_adapter(
...
-            method = abi.Method.from_signature(abi_method_signature)
-            selector = method.get_selector()
-            abi_argument_types = [a.type for a in method.args]
-            if method.returns.type != abi.Returns.VOID:
-                abi_return_type = method.returns.type
-
-        return is_app, abi_argument_types, abi_return_type, method, selector
-
-
-class ABIContractExecutor:
```
Comment (tzaffi) [inline graviton/blackbox.py:577]: This logic now lives in (a new form) inside `graviton/abi_strategy.py`

File: setup.py [hunk 2]
```diff
@@ -12,7 +12,10 @@
     author="Algorand",
     author_email="pypiservice@algorand.com",
     python_requires=">=3.9",
-    install_requires=["py-algorand-sdk>=2.0.0", "tabulate==0.9.0"],
+    install_requires=[
+        "py-algorand-sdk@git+https://github.com/algorand/py-algorand-sdk@better-exports",
+        "tabulate==0.9.0",
+    ],
```
Comment (tzaffi) [inline setup.py:18]: ^^^ TODO: revert to correct version before merging ^^^

File: tests/integration/abi_router_test.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
+from dataclasses import dataclass
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:1]: New test-file broken off from `tests/integration/abi_router_test.py` but with the same tests, only refactored to use `class Simulation` and `class ABIMethodCallStrategy`

File: tests/integration/abi_router_test.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+            return
+
+        method = method_runner.method
+        sim = Simulation(
+            algod,
+            ExecutionMode.Application,
+            method_runner.teal,
+            invariants,
+            abi_method_signature=good_abi_args.method_signature(),
+        )
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:282]: ## STEP 1. Prepare a `Simulation` object for executing dry runs, and asserting that it behaved as expected
It needs:
* an `algod`
* a `mode`
* `teal` program
* `invariants` which is a "predicates" dictionary mapping from `DryRunProperty`s to expressions that should evalute true for a dry run execution
* an `abi_method_signature` which will be used for validating inputs and decoding outputs

File: tests/integration/abi_router_test.py [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+            return
+
+        method = method_runner.method
+        sim = Simulation(
+            algod,
+            ExecutionMode.Application,
+            method_runner.teal,
+            invariants,
+            abi_method_signature=good_abi_args.method_signature(method),
+        )
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:281]: ## STEP 1. Prepare a `Simulation` object for executing dry runs while asserting that it behaved as expected
It needs:
* an `algod`
* a `mode`
* `teal` program
* `invariants` which is a "predicates" dictionary mapping from `DryRunProperty`s to expressions that should evalute true for a dry run execution
* an `abi_method_signature` which will be used for validating inputs and decoding outputs

File: tests/integration/abi_router_test.py [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+ALL_CASES = QUESTIONABLE_CASES + YACC_CASES
+
+
+def get_aa_strat(method_runner, abi_args_mod=None) -> ABIMethodCallStrategy:
+    return ABIMethodCallStrategy(
+        method_runner.contract,
+        RandomABIStrategyHalfSized,
+        num_dryruns=NUM_ROUTER_DRYRUNS,
+        abi_args_mod=abi_args_mod,
+    )
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:248]: ## STEP 2. Prepare a `ABIMethodCallStrategy` object for generating inputs for the dry-run according to a particular strategy.
It needs:
* an A.B.I. JSON `contract`
* an `abi_strategy` _type_
* the number of dry-runs to generate (`num_dryruns`)
* an `abi_args_mod` indicating how (if at all) to modify the inputs in an effort to test edge cases (`abi_args_mod=None` indicates that we should just the the generate inputs without further modifications)

File: tests/integration/abi_router_test.py [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+
+        for is_app_create, on_complete in call_types:
+            sim_result = sim.run_and_assert(
+                good_abi_args,
+                txn_params=TxParams.for_app(
+                    is_app_create=is_app_create,
+                    on_complete=on_complete,
+                ),
+                msg=msg(),
+            )
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:299]: ## STEP 3. Call the Simulations's `run_and_assert()` method
It needs:
1. the `abi_args_strategy` created in previous step (2)
2. `txn_params` which will need the `is_app_create` and `on_complete` parameters which indicate what the dry-run's call context is

File: tests/integration/abi_router_test.py [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+ALL_CASES = QUESTIONABLE_CASES + YACC_CASES
+
+
+def get_aa_strat(method_runner, abi_args_mod=None) -> ABICallStrategy:
+    return ABICallStrategy(
+        method_runner.contract,
+        RandomABIStrategyHalfSized,
+        num_dryruns=NUM_ROUTER_DRYRUNS,
+        abi_args_mod=abi_args_mod,
+    )
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:248]: ## STEP 2. Provide an `ABICallStrategy` generating dry runs inputs for each call
It needs:
* an ABI JSON `contract`
* an `ABIStrategy` _type_ to generate each individual arg
* `num_dryruns` for the number of input runs to generate
* `abi_args_mod` - possibly an `ABIArgsMod` **enum** that specifies how to _mutate_ the tweak for exploring what happens outside of the boundaries of anticipated inputs

File: tests/integration/abi_router_test.py [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+
+        for is_app_create, on_complete in call_types:
+            sim_result = sim.run_and_assert(
+                good_abi_args,
+                txn_params=TxParams.for_app(
+                    is_app_create=is_app_create,
+                    on_complete=on_complete,
+                ),
+                msg=msg(),
+            )
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:299]: ## STEP 3. Call the Simulations's `run_and_assert()` method
It needs:
* the `abi_args_strategy` created in previous step (2)
* `txn_params` which will need the `is_app_create` and `on_complete` parameters which indicate what the dry-run's call context is
* `msg` - an optional message to print to help discover failing test needles in the haystack

File: tests/integration/abi_test.py [hunk 10]
```diff
@@ -234,480 +228,3 @@ def test_roundtrip_abi_strategy(roundtrip_app):
     assert expected_mut == mut, inspector.report(
         args, "expected_mut v. mut", last_steps=last_rows
     )
-
-
-# ---- ABI Router Dry Run Testing - SETUP ---- #
```
Comment (ahangsu) [inline tests/integration/abi_test.py:239]: looks like these parts goes to https://github.com/algorand/pyteal/pull/634/files

File: graviton/inspector.py [hunk 11]
```diff
@@ -504,7 +504,7 @@ def budget_consumed(self) -> Optional[int]:
         """
         return self.dig(DRProp.budgetConsumed) if self.is_app() else None
 
-    def last_log(self) -> Optional[str]:
+    def last_log(self) -> Any:
```
Comment (ahangsu) [inline graviton/inspector.py:507]: the change in type is because `dig` is returning `Any`, right?

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,6 +1,14 @@
...
 # Changelog
 
+## `v0.9.0` (_aka_ 🐐)
+
+### Breaking changes
+
+* `class ABIContractExecutor` renamed to `ABICallStrategy` and moved from `graviton/blackbox.py` to `graviton/abi_strategy.py`
+
+### Added
+* `class Simulation` in `graviton/sim.y` unifies the ability to run an argument strategy and check that invariants hold using its `run_and_assert()` method
```
Comment (ahangsu) [inline CHANGELOG.md:11]: nit: typo fix
```suggestion
* `class Simulation` in `graviton/sim.py` unifies the ability to run an argument strategy and check that invariants hold using its `run_and_assert()` method
```

File: graviton/inspector.py (hunk 11 above)
Comment (tzaffi) [reply graviton/inspector.py:507]: Yes!

File: graviton/blackbox.py [hunk 13]
```diff
@@ -303,6 +301,11 @@ def asdict(self, drop_nones: bool = True) -> Dict[str, Any]:
 
         return {k: v for k, v in d.items() if v is not None}
 
+    def update(self, other: "DryRunTransactionParams") -> None:
+        for k, v in vars(other).items():
+            if v is not None:
+                setattr(self, k, v)
+
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply graviton/blackbox.py:308]: Hmmm... I do believe this is a reference in 634:

https://github.com/algorand/pyteal/blob/2dafd5213adfcca4ad97aebf06f0ca7002c17d85/tests/blackbox.py#L767

File: graviton/abi_strategy.py [hunk 14]
```diff
@@ -183,3 +204,161 @@ def get(self) -> PyTypes:
...
+            handle adding the method selector so this param.
+            But if set False: when providing `inputs`
+            ensure that the 0'th argument for method calls is the selector.
+            And when set True: when NOT providing `inputs`, the selector arg
+            at index 0 will be added automatically.
+
+        abi_args_mod (optional) - when desiring to mutate the args, provide an ABIArgsMod value
+        """
+        self.contract: abi.Contract = abi.Contract.from_json(contract)
+        self.argument_strategy: Optional[Type[ABIStrategy]] = argument_strategy
```
Comment (ahangsu) [inline graviton/abi_strategy.py:253]: nit: I am a bit unconvinced about why `self.argument_strategy` should be optional, for we are assigning from argument `argument_strategy` of `Type[ABIStrategy]`.

File: graviton/abi_strategy.py [hunk 15]
```diff
@@ -183,3 +204,161 @@ def get(self) -> PyTypes:
...
+            # TODO: when incorporating hypothesis strategies, we'll need a more holistic
+            # approach that looks at relationships amongst various args
+            args = tuple(
+                selector_mod(prefix) + [self.get(atype) for atype in arg_types]
+            )
+            return args_mod(args)
+
+        return [gen_args() for _ in range(self.num_dryruns)]
+
+    def get(self, gen_type: abi.ABIType) -> PyTypes:
```
Comment (ahangsu) [inline graviton/abi_strategy.py:363]: nit: I didn't find usage of this method in this pr, so I would suggest taking it out. If it is (or will be) used in elsewhere, I suggest renaming to `get_abi_arg_by_strategy`, or something more specific.

File: graviton/abi_strategy.py (hunk 14 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:253]: Good catch!

File: graviton/abi_strategy.py [hunk 16]
```diff
@@ -183,3 +204,161 @@ def get(self) -> PyTypes:
...
+        """
+        if method is None:
+            return []
+
+        return [cast(abi.ABIType, arg.type) for arg in self.abi_method(method).args]
+
+    def num_args(self, method: Optional[str]) -> int:
+        return len(self.argument_types(method))
+
+    def generate(self, method: Optional[str]) -> List[Sequence[PyTypes]]:
```
Comment (ahangsu) [inline graviton/abi_strategy.py:287]: nit: I would suggest renaming to `generate_inputs` to align with previous implementation in `blackbox.py`, or renaming to `generate_inputs_for_method` to be specific.

File: graviton/blackbox.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ahangsu) [reply graviton/blackbox.py:308]: I see, so we do need to keep this method. A few other suggestions here:
- a bit worried about using `vars`, probably use https://stackoverflow.com/a/70168093 ?
- suggest renaming as `update_against_params`, for we are taking non empty fields in another instance and apply to self.

File: graviton/inspector.py (hunk 11 above)
Comment (ahangsu) [reply graviton/inspector.py:507]: okok, I think I am fine with the move, for no behavior change behind the curtain.

File: graviton/sim.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+
+@dataclass(frozen=True)
+class SimulationResults:
+    succeeded: bool
+    simulate_inspectors: List[DryRunInspector]
+    identities_inspectors: Optional[List[DryRunInspector]] = None
+
+
+class Simulation:
```
Comment (ahangsu) [inline graviton/sim.py:24]: I guess this class serves as the purpose of `ABIContractExecutor.dry_run_on_sequence`, together with some assertions in predicates?

File: graviton/sim.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+    def run_and_assert(
+        self,
+        inputs: InputStrategy,
+        *,
+        txn_params: Optional[TxParams],
+        verbose: bool = False,
+        msg: str = "",
+    ) -> SimulationResults:
+        assert inputs, "must provide actual inputs to run against!"
```
Comment (ahangsu) [inline graviton/sim.py:71]: nit: maybe we can lift this assertion by `inputs: InputStrategy`?

File: graviton/blackbox.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tzaffi) [reply graviton/blackbox.py:308]: * Regarding first bullet: this is a good point. But since I need both keys and values, I'll use the `asdict()` method of `dataclasses` instead
* Regarding second bullet: how about `update_fields` ?

[the remark this answers is not in the document]
Comment (ahangsu) [reply graviton/blackbox.py:308]: sounds good to me!

File: graviton/abi_strategy.py (hunk 15 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:363]: This is actually being used inside `generate()` in two list comprehensions in the local functions `args_mod()` and `gen_args()`.

In terms of naming. You bring up a good point. I think the current naming is confusing and vague. In particular, the _example generator_ methods:
* `generate` produces a `list[Sequence[PyTypes]]` which is supposed to be a list of _args-tuples_ with each _args-tuple_ intended as the input of a dry run
* `get` produces a single `PyTypes` 

One possible renaming would be:
* ~`generate()`~ $\rarr$ `generate_inputs()`
* ~`get()`~ $\rarr$ `generate_value()`

But I'm not totally convinced and am open to other suggestions.

File: graviton/abi_strategy.py (hunk 16 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:287]: woah, I made my comment above _without_ reading this first. JINX

File: graviton/sim.py (hunk 17 above)
Comment (tzaffi) [reply graviton/sim.py:24]: Yes, exactly. So basically:

`Simulation = (Teal logic for execution) + (predicates that must be satisfied)`

`run_and_assert: simulation + InputStrategy` &rarr; `SUCCESS or FAILURE`

In general, graviton needs a deep cleaning/update of it's documentation. I've just created #51 as a reminder

Comment (ahangsu) [reply graviton/sim.py:24]: sounds good, I was keeping track of deleted `ABIContractExecutor` and its methods, trying to find where are the new substitutions. So far I think these mechanisms have their replacements in place.

File: graviton/sim.py (hunk 18 above)
Comment (tzaffi) [reply graviton/sim.py:71]: I'm not sure I understand. In terms of strong typing, this assertion is unnecessary. But since the method is public I still want to ensure that at runtime the user provides actual inputs. But I concede that we could try and make even stronger assertions. For example, on line 85 `cast` is called to make mypy happy, but one could try to make some run-time assertions and validate the shape of `inputs`. I'm counting on the other executor validations would get kicked off on line 89 and hopefully providing enough actionable errors to help a user fix their problems.

File: graviton/abi_strategy.py (hunk 15 above)
Comment (ahangsu) [reply graviton/abi_strategy.py:363]: The possible renaming you proposed above makes sense to me.

File: graviton/sim.py (hunk 18 above)
Comment (ahangsu) [reply graviton/sim.py:71]: I see, that makes sense for exported method. Yea type annotation doesn't help too much at runtime tho.

File: tests/unit/encode_test.py [hunk 19]
```diff
@@ -92,186 +90,6 @@ def test_encode_abi():
     )
 
 
-NONSENSE = "not a valid signature"
```
Comment (tzaffi) [inline tests/unit/encode_test.py:86]: Moved to new test file `executor_test.py`

File: graviton/blackbox.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tzaffi) [reply graviton/blackbox.py]: Addressed in https://github.com/algorand/graviton/pull/49/commits/d8541f210844dd34c6ffbe7d2533b7945c1b9fe8 + https://github.com/algorand/graviton/commit/809d54466f14bb7c67cd5989b6b9dbec35dbcfc5

File: graviton/abi_strategy.py (hunk 14 above)
Comment (tzaffi) [reply graviton/abi_strategy.py]: cf. https://github.com/algorand/graviton/pull/49/commits/dfc2f7b0730006bc53f8dcd3013b92b7f8dd5052

File: graviton/abi_strategy.py (hunk 15 above)
Comment (tzaffi) [reply graviton/abi_strategy.py]: cf https://github.com/algorand/graviton/pull/49/commits/dfc2f7b0730006bc53f8dcd3013b92b7f8dd5052

File: CHANGELOG.md [hunk 20]
```diff
@@ -1,6 +1,14 @@
 <!-- markdownlint-disable MD024 -->
 # Changelog
 
+## `v0.9.0` (_aka_ 🐐)
+
+### Breaking changes
+
+* `class ABIContractExecutor` renamed to `ABICallStrategy` and moved from `graviton/blackbox.py` to `graviton/abi_strategy.py`
```
Comment (tzaffi) [inline CHANGELOG.md:8]:
```suggestion
* `class ABIContractExecutor` renamed to `ABICallStrategy` and moved from `graviton/blackbox.py` to `graviton/abi_strategy.py`. Some of the methods have been renamed as well.
```

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,6 +1,14 @@
...
+
+### Added
+* `class Simulation` in `graviton/sim.py` unifies the ability to run an argument strategy and check that invariants hold using its `run_and_assert()` method
```
Comment (tzaffi) [inline CHANGELOG.md:12]:
```suggestion
* `class Simulation` in `graviton/sim.py` unifies the ability to run an argument strategy and check that invariants hold using its `run_and_assert()` method
* `class DryRunTransactionParameters` has a new method `update_fields()`

```

**Review by tzaffi — COMMENTED** [review]
wip

File: graviton/abi_strategy.py [hunk 22]
```diff
@@ -183,3 +204,162 @@ def get(self) -> PyTypes:
...
+
+        modify_selector = False
+        if (action := self.abi_args_mod) in (
+            ABIArgsMod.selector_byte_delete,
+            ABIArgsMod.selector_byte_insert,
+            ABIArgsMod.selector_byte_replace,
+        ):
+            assert (
+                prefix
+            ), f"{self.abi_args_mod=} which means we need to modify the selector, but we don't have one available to modify"
```
Comment (ahangsu) [inline graviton/abi_strategy.py:316]: nit: just reading this line of assertion, and thinking about a suggestion:
- add a field representing `no_mutation` in `ABIArgsMod`
- then on initialization assignment: `self.abi_args_mod = abi_args_mod or ABIArgsMod.no_mutation`

We can shrink the type `self.abi_args_mod: ABIArgsMod | None` to `ABIArgsMod`.

File: graviton/abi_strategy.py [hunk 23]
```diff
@@ -183,3 +204,162 @@ def get(self) -> PyTypes:
...
+            modify_selector = True
+
+        def selector_mod(prefix):
```
Comment (ahangsu) [inline graviton/abi_strategy.py:319]: nit: suggested renaming `selector_modification`

File: graviton/abi_strategy.py [hunk 24]
```diff
@@ -183,3 +204,162 @@ def get(self) -> PyTypes:
...
+                ), f"expected action={ABIArgsMod.selector_byte_replace} but got [{action}]"
+                idx = random.randint(0, 3)
+                selector = (
+                    selector[:idx]
+                    + bytes([(selector[idx] + 1) % 256])
+                    + selector[idx + 1 :]
+                )
+            return [selector]
+
+        def args_mod(args):
```
Comment (ahangsu) [inline graviton/abi_strategy.py:343]: nit: suggested renaming `args_modification`

**Review by ahangsu — COMMENTED** [review]
I think `abi_strategy.py` looks neat, looking into `sim.py` and related test. Hopefully we can wrap this up this week

File: graviton/sim.py [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+        self.simulate_dre: DryRunExecutor = DryRunExecutor(
+            algod,
+            mode,
+            simulate_teal,
+            abi_method_signature=abi_method_signature,
+            omit_method_selector=omit_method_selector,
+            validation=validation,
+        )
+        self.identities_dre: Optional[DryRunExecutor] = None
+        if identities_teal:
```
Comment (ahangsu) [inline graviton/sim.py:53]: I might remember or understand it wrong, is this for cross version check (i.e.., ver6 vs v8)?

**Review by ahangsu — COMMENTED** [review]
I finished my pass on `sim.py`, should be almost good to go, some TODOs in code (mostly comments and documentation) need to be done.

File: CHANGELOG.md [hunk 26]
```diff
@@ -3,6 +3,10 @@
 
 ## `v0.9.0` (_aka_ 🐐)
 
+### Bugs fixed
+
+`go-algorand` PR # ??? has changed the way errors are handled during dry which broke the methods `error()` and `error_message` of `DryRunInspector`.  This is fixed in [#49](https://github.com/algorand/graviton/pull/49)
```
Comment (tzaffi) [inline CHANGELOG.md:8]: TODO: Don't merge without providing this information

File: tests/integration/abi_router_test.py [hunk 27]
```diff
@@ -308,10 +308,9 @@ def msg():
     err\ opcode                                 # pyteal generated err's ok
 |   assert\ failed\ pc=                         # pyteal generated assert's ok
 |   invalid\ ApplicationArgs\ index             # failing because an app arg wasn't provided
-|   extract\ range\ beyond\ length\ of\ string  # failing because couldn't extract when omitted final arg or jammed in tuple
+|   extraction\ end\ 16\ is\ beyond\ length     # failing because couldn't extract when omitted final arg or jammed in tuple
```
Comment (tzaffi) [inline tests/integration/abi_router_test.py:311]: error messages have changed

File: .github/workflows/build.yml [hunk 28]
```diff
@@ -46,7 +46,7 @@ jobs:
           echo $installed
           [[ $installed =~ "Python ${expected}" ]] && echo "Configured Python" || (echo "Failed to configure Python" && exit 1)
       - name: Start algod
-        run: make algod-start
+        run: make algod-start-report
```
Comment (tzaffi) [inline .github/workflows/build.yml:49]: adding a printout of the commit-hash when C.I. stands up the algod.

File: graviton/abi_strategy.py (hunk 22 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:338]: This is a very good suggestion. I'm trying to finish up work on the PR and I got stuck on some extra curve balls. So will as to punt as a future improvement.

File: graviton/abi_strategy.py (hunk 23 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:341]: This is an understandable nit. However, I would have to change the usage of "mod" in several places including in pyteal. So I politely push back but am open to further discussion.

File: graviton/abi_strategy.py (hunk 24 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:365]: see my previous comment

File: graviton/abi_strategy.py (hunk 23 above)
Comment (ahangsu) [reply graviton/abi_strategy.py:341]: sound fine with me, adding a line of comment should be suffice for me, same applies to comment below

File: graviton/sim.py (hunk 25 above)
Comment (ahangsu) [reply graviton/sim.py:53]: 👍

File: graviton/abi_strategy.py (hunk 23 above)
Comment (ahangsu) [reply graviton/abi_strategy.py:341]: lgtm

File: graviton/abi_strategy.py (hunk 22 above)
Comment (ahangsu) [reply graviton/abi_strategy.py:338]: got it, I think prioritizing other curve balls makes sense, I leave this thread open, but we probably do not want to work on it at this moment.

Comment (ahangsu) [reply graviton/abi_strategy.py:338]: (for functionality of this method is not lost)

File: .github/workflows/build.yml (hunk 28 above)
Comment (ahangsu) [reply .github/workflows/build.yml:49]: you want to know the version code to locate some dryrun related behavior difference?

File: setup.py (hunk 2 above)
Comment (tzaffi) [reply setup.py:18]: @ahangsu - I'm giving up on this for now. But I created #52 so it gets done ASAP.

Comment (ahangsu) [reply setup.py:18]: 👍 I am fine either way, and releasing sdk is beyond us. We can merge as is and pin the right version after sdk release, is that fine with you?

Comment (tzaffi) [reply setup.py:18]: Agreed and resolving this confo

File: graviton/abi_strategy.py [hunk 29]
```diff
@@ -4,16 +4,28 @@
...
 
+
 from graviton.models import PyTypes
 
 
 class ABIStrategy(ABC):
+    """
+    TODO: when incorporating hypothesis strategies, we'll need a more holistic
+    approach that looks at relationships amongst various args.
+    Current approach only looks at each argument as a completely independent entity.
```
Comment (ahangsu) [inline graviton/abi_strategy.py:22]: no action needed, I see hypothesis cross the PR, together with TODO, are you referring to hypothesis https://hypothesis.readthedocs.io/en/latest/data.html?

**Review by ahangsu — COMMENTED** [review]
I am leaning towards approving and merging, just one question on hypothesis and related TODO, but I don't expect we will take any action here (for it would be non trivial).

File: .github/workflows/build.yml (hunk 28 above)
Comment (tzaffi) [reply .github/workflows/build.yml:49]: Yes. That's how I was able to understand that latest graviton test failures were because I was running an old version of algod. In fact, I ended up needing to run `docker compose pull` to fix the discrepancy.

File: graviton/abi_strategy.py (hunk 29 above)
Comment (tzaffi) [reply graviton/abi_strategy.py:22]: Yes. Here's a nice video intro about it: https://www.youtube.com/watch?v=mkgd9iOiICc&ab_channel=ArjanCodes

**Review by ahangsu — APPROVED** [review]
